### PR TITLE
UI: DAG Calendar View

### PR DESCRIPTION
**Pull request by BenoitHanotte**
# Motivation

Introduce a DAG Calendar View to provide visibility over the full state of the dag by displaying the aggregated dag runs' states in a calendar.
This makes it possible to monitor the state of thousands of dag runs in a single view that is concise and easy to understand. It is particularly useful to monitor the state of large backfills:

image

Each day is displayed with a color according to the dag runs' states for that day:
- If at least one dag run has failed for a day, the day will be displayed as "failed".
- If all dag runs have succeeded, the day will be shown as "succeeded". 
- If there are still running dag runs (and no failed dag run) for that day, the day will be shown as "running".

A tooltip, for each day, gives the exact number of dag runs in each state for the day.
image

Clicking on a day redirects to the tree view for that day to show the task instances for that day.

# Implementation
- New `/calendar` endpoint in `views.py`. 
This view returns, for each day, the number of dag run in each state. Thus, only the `dag_run` table is queried, and the query returns at most 3 * days records. Thus, this view puts little strain on the database, even for dags with a large number of task instances.
- New `calendar.html` template, `calendar.js` and `calendar.css` static files.
The calendar view is implemented usng d3, thus no additional dependency has been introduced.

The proposal has been discussed on the dev mailing list:
https://lists.apache.org/thread.html/r6f2ea3f83d268a665ac63219553718ed089a01e9530627cca0572ab1%40%3Cdev.airflow.apache.org%3E

### CODE REVIEW

File: airflow/www/static/js/calendar.js [hunk 1]
```diff
@@ -0,0 +1,269 @@
+/* eslint-disable func-names */
+/* eslint-disable no-underscore-dangle */
```
Comment (bbovenzi) [inline airflow/www/static/js/calendar.js]: We should remove both `eslint-disable`. Neither rule is too cumbersome for this file.

File: airflow/www/static/css/calendar.css [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+rect.running {
+  fill: lime;
+}
+
+rect.failed {
+  fill: red;
+}
+
+rect.no_status {
+  fill: #eee;
```
Comment (bbovenzi) [inline airflow/www/static/css/calendar.css]: No Status is white elsewhere in the app. We should stay consistent here too.

**Review by bbovenzi — CHANGES_REQUESTED** [review]
Technical: Overall looks good. In addition to the comments I left, the `calendar.js` file needs a linting pass. There were too many little issues to comment on.

General: I still feel like we could make this view more data-rich. It would be great to do a heatmap of the ratio of successful and failed runs. After that, an indication of number of total runs a day. It may take some trial & error to get it both of those to look right within the same chart.

File: airflow/www/static/css/calendar.css (hunk 2 above)
Comment (BenoitHanotte) [reply airflow/www/static/css/calendar.css]: Thanks, I have fixed it

File: airflow/www/static/js/calendar.js (hunk 1 above)
Comment (BenoitHanotte) [reply airflow/www/static/js/calendar.js]: I removed the second one, but disabling `func-names` is required to be able to use d3-tip as:
```
      .on('mouseover', function (data) {
        const tt = tipHtml(data);
        dayTip.direction('n');
        dayTip.show(tt, this);
      })
      .on('mouseout', function (data) {
        dayTip.hide(data, this);
      });
```
Using arrow-style functions won't work for some reason.
Removing `func-names` requires extracting the anonymous functions, but that also makes the code a bit heavier. I believe this is the reason `func-names` is currently disabled in `tree.js` as well.

What do you think? should we keep it this way or should we extract the tooltip functions?

**Comment by BenoitHanotte** [comment]
> Technical: Overall looks good. In addition to the comments I left, the `calendar.js` file needs a linting pass. There were too many little issues to comment on.
> 
> General: I still feel like we could make this view more data-rich. It would be great to do a heatmap of the ratio of successful and failed runs. After that, an indication of number of total runs a day. It may take some trial & error to get it both of those to look right within the same chart.

Thanks for the review, I applied `eslint` and addressed the comments.

I also made a change to use a heatmap to fill the day status, here is how it currently looks like:
image
If there is at least one failure for a day, the color will be a mix of green and red according to the ratio of failed runs over the number of runs for that day (https://github.com/apache/airflow/pull/15423/files#diff-4bd7c1010027099f09d8d824b871bfc65c54843dc5283d3ec336182c623712b6R216)
On the screenshot above you can see 3 days with faile DAG runs: Jan 16 has 1 failed run, Jan 17 has 4 and Jan 18 has 8 failed DAG runs.
Don't hesitate to let me know what you think! :)

File: airflow/www/static/js/calendar.js (hunk 1 above)
Comment (uranusjr) [reply airflow/www/static/js/calendar.js]: I’m far from fluent in JavaScript, but would sound like this pass?

```js
.on('mouseout, function hideDayTipOnMouseOut(data) {
  dayTip.hide(data, this)
});
```

Comment (BenoitHanotte) [reply airflow/www/static/js/calendar.js]: @uranusjr thanks, `eslint` seems to be ok with that, I have updated the code and removed the eslint exception :+1:

**Comment by bbovenzi** [comment]
The red-green mixing doesn't seem right.  I don't know if we just have to play with the function you already have or use a different one like this: https://www.geeksforgeeks.org/d3-js-interpolaterdylgn-function/

Code-wise things look good to me. I think we still need to refine the UI. I'll pull it down again and experiment a bit.

**Comment by BenoitHanotte** [comment]
> The red-green mixing doesn't seem right. I don't know if we just have to play with the function you already have or use a different one like this: https://www.geeksforgeeks.org/d3-js-interpolaterdylgn-function/
> 
> Code-wise things look good to me. I think we still need to refine the UI. I'll pull it down again and experiment a bit.

The interpolation function `interpolaterdylgn` does not interpolate between green and red but between 3 colors: green, yellow and red. We can't hardcode these colors in the UI as the backend is responsible for providing the status colors to the UI.
I can however easily change the interpolation function from `interpolateRgb` to `interpolateHsl` which does the interpolation in the HSL space and thus goes from green to red through yellow. It gives the following colors:
image

However, yellow is already used as the color for the `up_for_retry` and `upstream_failed` task instance statuses. Displaying a day as yellow might be misleading to users as they might think there is a just task that will be retried that day?

**Comment by BenoitHanotte** [comment]
@bbovenzi I have pushed the change to use `interpolateHsl` (so that we go from green to read going through yellow), and raised the interpolation floor to 0.5 so that days with just a few failures don't appear in a green that is close to the "success" green and would thus "hide" the failures.
The current result is the following:
image

Let me know what you think :)

**Comment by bbovenzi** [comment]
Way better!  Good call on `interpolateHsl` We should update the legend to reflect this which should help mitigate your first concern.

**Comment by ryanahamilton** [comment]
Would it be more useful to reverse the year ordering? Seems like the current year is likely to be the most relevant to the user.

**Comment by BenoitHanotte** [comment]
> Way better! Good call on `interpolateHsl` We should update the legend to explain the range of colors. That should help mitigate your first concern.

I have updated the legend, it adds a bit of complexity as to display the "gradient" in the legend it required moving the legend in the svg object:
image

> Would it be more useful to reverse the year ordering? Seems like the current year is likely to be the most relevant to the user.

I experimented with it might be a bit confusing as consecutive days are not juxtaposed from left-to-right and top-to-bottom anymore:
image

Let me know which option you prefer.

File: airflow/www/views.py [hunk 3]
```diff
@@ -2078,6 +2078,71 @@ def tree(self):
             external_log_name=external_log_name,
         )
 
+    @expose('/calendar')
+    @auth.has_access(
+        [
+            (permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG),
+            (permissions.ACTION_CAN_READ, permissions.RESOURCE_TASK_INSTANCE),
+            (permissions.ACTION_CAN_READ, permissions.RESOURCE_TASK_LOG),
```
Comment (ashb) [inline airflow/www/views.py]:
```suggestion
```

I don't think this makes sense to include, as it doesn't directly read the task logs.

**Comment by bbovenzi** [comment]
The gradient legend is perfect. 
I think the original top-to-bottom arrangement was better.

File: airflow/www/templates/airflow/calendar.html [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+{% extends "airflow/dag.html" %}
+{% block page_title %}{{ dag.dag_id }} - Tree - {{ appbuilder.app_name }}{% endblock %}
+
+{% block head_css %}
+  {{ super() }}
+  <meta name="dag_id" content="{{ dag.dag_id }}">
+  <meta name="tree_url" content="{{ url_for('Airflow.tree') }}">
+  <link rel="stylesheet" type="text/css" href="{{ url_for_asset('calendar.css') }}">
+  </style>
```
Comment (bbovenzi) [inline airflow/www/templates/airflow/calendar.html]:
```suggestion
```

File: airflow/www/templates/airflow/calendar.html [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+{% endblock %}
+
+{% block content %}
+  {{ super() }}
+  <hr>
+  <div id="svg_container">
+    <img id='loading' width="50" src="{{ url_for('static', filename='loading.gif') }}">
+    <svg id="calendar-svg">
+    </svg>
```
Comment (bbovenzi) [inline airflow/www/templates/airflow/calendar.html]:
```suggestion
    <svg id="calendar-svg"></svg>
```

File: airflow/www/templates/airflow/calendar.html [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  {{ super() }}
+  <script src="{{ url_for_asset('d3.min.js') }}"></script>
+  <script src="{{ url_for_asset('d3-tip.js') }}"></script>
+  <script src="{{ url_for_asset('calendar.js') }}"></script>
+  <script>
+    const statesColors = {};
+    statesColors["no_status"] = "white";
+    {% for state in ['failed', 'success', 'running'] %}
+    statesColors["{{state}}"] = "{{state_color_mapping[state]}}";
+    {% endfor %}
```
Comment (bbovenzi) [inline airflow/www/templates/airflow/calendar.html]:
```suggestion
    {% for state in ['failed', 'success', 'running'] %}
      statesColors["{{state}}"] = "{{state_color_mapping[state]}}";
    {% endfor %}
```

**Review by bbovenzi — CHANGES_REQUESTED** [review]
Looking great! I only added some code styling changes from a frontend perspective.

**Review by ashb — APPROVED** [review]

**Comment by ashb** [comment]
@bbovenzi Are you happy with this PR now?

**Review by bbovenzi — APPROVED** [review]
Yep. LGTM.

**Pull request merged**